lib/indexer/indexer.js

```diff
@@ -0,0 +1,812 @@
+      } catch (e) {
+        this.emit('error', e);
+      }
+    });
+
+    this.client.on('block disconnect', async (entry, block, view) => {
+      if (this.rescanning)
+        return;
+      try {
+        await this.removeBlock(entry, block, view);
+      } catch (e) {
+        this.emit('error', e);
+      }
+    });
+
+    this.client.on('block rescan', async (entry, block, view) => {
+      try {
+        await this.rescanBlock(entry, block, view);
+      } catch (e) {
+        this.emit('error', e);
+      }
+    });
+
+    this.client.on('chain reset', async (tip) => {
+      try {
+        await this.resetChain(tip);
+      } catch (e) {
+        this.emit('error', e);
+      }
+    });
+
+    this.client.on('chain tip', async () => {
+      this.logger.debug('Indexer: finished rescan');
+      const tip = await this.getTip();
+      this.emit('chain tip', tip);
+    });
+  }
+
+  /**
+   * Ensure prefix directory (prefix/index).
+   * @returns {Promise}
+   */
+
+  async ensure() {
+    if (fs.unsupported)
+      return undefined;
+
+    if (this.options.memory)
+      return undefined;
+
+    return fs.mkdirp(this.options.prefix);
+  }
+
+  /**
+   * Open the indexdb, wait for the database to load.
+   * @returns {Promise}
+   */
+
+  async open() {
+    await this.ensure();
+    await this.db.open();
+    await this.db.verify(layout.V.encode(), 'index', 0);
+
+    await this.verifyNetwork();
+
+    await this.connect();
+  }
+
+  /**
+   * Verify network.
+   * @returns {Promise}
+   */
+
+  async verifyNetwork() {
+    const raw = await this.db.get(layout.O.encode());
+
+    if (!raw) {
+      const b = this.db.batch();
+      b.put(layout.O.encode(), fromU32(this.network.magic));
+      return b.write();
+    }
+
+    const magic = raw.readUInt32LE(0, true);
+
+    if (magic !== this.network.magic)
+      throw new Error('Network mismatch for Indexer.');
+
+    return undefined;
+  }
+
+  /**
+   * Close the indexdb, wait for the database to close.
+   * @returns {Promise}
+   */
+
+  async close() {
+    await this.disconnect();
+    return this.db.close();
+  }
+
+  /**
+   * Connect to the chain server (client required).
+   * @returns {Promise}
+   */
+
+  async connect() {
+    return this.client.open();
+  }
+
+  /**
+   * Disconnect from chain server (client required).
+   * @returns {Promise}
+   */
+
+  async disconnect() {
+    return this.client.close();
+  }
+
+  /**
+   * Sync state with server on every connect.
+   * @returns {Promise}
+   */
+
+  async syncNode() {
+    const unlock = await this.lock.lock();
+    try {
+      this.logger.info('Resyncing from server...');
+      await this.syncState();
+      await this.syncChain();
+    } finally {
+      unlock();
+    }
+  }
+
+  /**
+   * Initialize and write initial sync state.
+   * @returns {Promise}
+   */
+
+  async syncState() {
+    const cache = await this.getState();
+
+    if (cache) {
+      this.state = cache;
+      this.height = cache.height;
```

Comment:
I don't think this will ever have an effect. the `ChainState` class used for the indexer doesn't serialize the height in `toRaw` which means it never gets saved in the database which means it never gets retrieved in `getState` above. This is the same with `state.startHeight`. This means you could have a saved state but none of the relevant information for the indexer sync.

Reply:
It looks like `height` is serialized in `toRaw` of `ChainState`, see: https://github.com/bcoin-org/bcoin/pull/424/files/4decdcfcd47575cc8e26add040d9a086169c8700#diff-024972a051090cc189bfa44111d82aa5R75

Reply:
Ah yeah, you're right. Not sure why that wasn't coming through correctly. I might've been on some older commit. Related though, it might be useful to expose the ChainState in records as I'm finding it useful when working with custom indexers (now you have to copy and paste the implementation in order to use ChainState in an indexer, which is where my mistake was coming from). 

Sorry for the false flag